### DISCUSSION TITLE
Custom remove function wrapper to ensure that all orphanRemoval-relations are populated

### DISCUSSION THREAD

**Discussion by Jcalcaldev** [mikro-orm/mikro-orm#5298] [Ideas]
**Is your feature request related to a problem? Please describe.**

On another thread I was made to realize that orphanRemoval doesn't work if relations are not populated, this might be a problem and devs might forget to populate deeply nested relations in some cases.

**Describe the solution you'd like**

I've already implemented a partial solution that seems to work. Basically, upon remove, it checks if all of the entitiy(es) to remove have orphanRemoval-marked relations and forces them to be populated or throws an error. It works on 1:m and 1:1 relations as far as I know.

Here's the proposed solution:

```ts
export class CustomEntityManager extends EntityManager {
  remove<Entity extends object>(
    toRemove: Entity | Reference<Entity> | Iterable<Entity | Reference<Entity>>
  ): this {
    this.validateOrphanRemovalsPopulated<Entity>(toRemove);

    return super.remove(toRemove);
  }

  async removeAndFlush(
    toRemove: AnyEntity | Reference<AnyEntity> | Iterable<AnyEntity | Reference<AnyEntity>>
  ): Promise<void> {
    this.validateOrphanRemovalsPopulated(toRemove);

    return await super.removeAndFlush(toRemove);
  }

  private validateOrphanRemovalsPopulated<Entity extends object>(
    toRemove: Entity | Reference<Entity> | Iterable<Entity | Reference<Entity>>
  ): void {
    const entities = Symbol.iterator in toRemove ? toRemove : [toRemove];

    for (const entity of entities) {
      const meta = this.getMetadata().get(entity.constructor.name);

      const orphanRemovablePropertiesNames = Object.values(meta.properties)
        .filter(prop => prop.kind !== 'scalar' && prop.orphanRemoval)
        .map(prop => prop.name);

      for (const propertyName of orphanRemovablePropertiesNames) {
        const subEntity: BaseEntity | Collection<BaseEntity> | undefined = (entity as AnyEntity)[
          propertyName
        ];

        if (subEntity) {
          if (!subEntity.isInitialized()) {
            throw new Error(
              `The relation ${propertyName} is not populated on entity ${meta.className}`
            );
          }

          if (subEntity instanceof Collection) {
            const collection = subEntity.getItems();
            this.validateOrphanRemovalsPopulated(collection);
          } else {
            this.validateOrphanRemovalsPopulated(subEntity);
          }
        }
      }
    }
  }
}
```

A possible improvement for the ORM would be for the remove to take an optional argument to run a check of these sorts behind the scenes.

**Describe alternatives you've considered**

Custom solution up above

**Additional context**

Nope

**Comment by B4nan** [comment]
I think it would make more sense to load the relation automatically, we could make that opt out and throw in such case, then it wont be breaking anyhow.

**Comment by Jcalcaldev** [comment]
Sounds good, any way to make sure that all orphanRemovals go as expected any time without requiring developers to remember every time what to populate, same as cascades would in the database level
